mc-sema/cfgToLLVM/x86Instrs.cpp

```diff
@@ -169,12 +169,144 @@ InstTransResult liftInstrImpl(InstPtr ip, BasicBlock *&block, NativeBlockPtr nb,
-    errs() << "Unsupported opcode " << opcode << "!";
+    errs() << "Unsupported opcode " << opcode << " ("
+      << IP->getOpcodeName(opcode) << "): ";
     IP->printInst( &inst, errs(), to_string<VA>(ip->get_loc(), hex));
-    // Print out the unhandled opcode.
-    errs() << to_string<VA>(ip->get_loc(), hex) << " ";
-    errs() << strOut.str() << "\n";
-    errs() << inst.getOpcode() << "\n";
+    errs() << "\n";
+
+    // In the case that we can't find the opcode, try building it out with
+    // inline assembly calls in LLVM instead. If at any point we don't know
+    // how to build the assembly, give up on trying to do so.
+    const MCRegisterInfo *MRI = natM->getMCContext()->getRegisterInfo();
+    const MCInstrInfo *MII = natM->getTarget()->createMCInstrInfo();
+
+    // Build the inline asm and the constraint string.
+    string inlineAsm = "", constraints = "";
+
+    // Use the printer to build the ASM string. We'll need to substitute
+    // $ for $$, though.
+    {
+      IP->printInst(&inst, strOut, "");
+      for (char c : strOut.str()) {
+        if (c == '$')
+          inlineAsm += "$$";
+        else
+          inlineAsm += c;
+      }
+    }
+
+    // Find all of the involved registers and build the LLVM
+    unsigned argnum = 0;
+    SmallVector<Value *, 3> operands;
+    SmallVector<Type *, 3> resultTypes;
+    for (unsigned i = 0; i < inst.getNumOperands(); i++) {
+      MCOperand &op = inst.getOperand(i);
+      if (op.isReg()) {
+        if (argnum++ > 0) constraints += ",";
+        if (i < MII->get(opcode).getNumDefs()) {
+          constraints += "=";
+
+          // Find the size of the register. Surprisingly, this is how you
+          // have to do that.
+          unsigned maxSize = 0;
+          for (auto rc = MRI->regclass_begin();
+              rc != MRI->regclass_end(); ++rc) {
+            if (rc->contains(op.getReg())) {
+              maxSize = std::max(maxSize, rc->getSize());
+            }
+          }
+
+          if (maxSize > 8) {
+            // LLVM can't handle register constraints of i128, so we
+            // need to map this to <16 x i8>.
+            resultTypes.push_back(VectorType::get(
+              Type::getInt8Ty(block->getContext()), maxSize));
+          } else if (maxSize > 0) {
+            resultTypes.push_back(IntegerType::get(block->getContext(),
+                  8 * maxSize));
+          } else {
+            errs() << "Can't find register size for " <<
+              MRI->getName(op.getReg()) << '\n';
+            goto failinlineasm;
+          }
+        } else {
+          Value *readReg = nullptr;
+          unsigned maxSize = 0;
+          for (auto rc = MRI->regclass_begin();
+              rc != MRI->regclass_end(); ++rc) {
+            if (rc->contains(op.getReg()))
+              maxSize = std::max(maxSize, rc->getSize());
+          }
+          switch (maxSize) {
+            case 1: readReg = ::R_READ<8>(block, op.getReg()); break;
```

Comment:
`R_READ` is implemented by `GENERIC_MC_READREG`, which takes the register size as an argument.